bookstore/archive.py

```diff
@@ -1,47 +1,89 @@
+"""Archival of notebooks"""
 import json
-import aiobotocore
-
 from asyncio import Lock
 from typing import Dict, NamedTuple
 
+import aiobotocore
 from notebook.services.contents.filemanager import FileContentsManager
 from tornado import ioloop
 
 from .bookstore_config import BookstoreSettings
-
 from .s3_paths import s3_key, s3_display_path
 
 
 class ArchiveRecord(NamedTuple):
+    """Representation of content to archive
+
+    A record to be archived is a tuple containing `filepath`, the `content` to
+    be archived, and the length of time in queue for archiving (`queued_time`).
+    """
     filepath: str
     content: str
     queued_time: float
 
 
 class BookstoreContentsArchiver(FileContentsManager):
+    """Archives notebooks to storage (S3) on save by UI or parameterized workflow
+
+    This class is a custom Jupyter `FileContentsManager` which holds information
+    on storage location, path to it, and file to be written there.
+    
+    Bookstore settings combine with the parent Jupyter application settings.
+    
+    A session is created for the current event loop. To write to a particular
+    path on S3, acquire a lock. After acquiring the lock, `archive`
+    authenticates using the storage service's credentials. If allowed, the
+    notebook is queued to be written to storage (i.e. S3).
+
+    Attributes
+    ----------
+    path_locks : dict
+        Dictionary of paths to storage and the lock associated with a path
+    path_lock_ready: ayncio mutex lock
+        A mutex lock associated with a path
+
+    See also
+    --------
+    https://jupyter-notebook.readthedocs.io/en/stable/extending/contents.html#contents-api
     """
-    Archives notebooks to S3 on save
-    """
-
     def __init__(self, *args, **kwargs):
         super(FileContentsManager, self).__init__(*args, **kwargs)
+
         # opt ourselves into being part of the Jupyter App that should have Bookstore Settings applied
         self.settings = BookstoreSettings(parent=self)
 
-        self.log.info("Archiving notebooks to {}".format(s3_display_path(
-            self.settings.s3_bucket, self.settings.workspace_prefix)))
+        self.log.info(
+            "Archiving notebooks to {}".format(
+                s3_display_path(self.settings.s3_bucket, self.settings.workspace_prefix)
+            )
+        )
 
-        self.session = aiobotocore.get_session()
+        # TODO: create a better check for credentials - keys and config or it will fail
```

Comment:
This reminds me we should better document how to use `minio` locally for development.